core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -368,35 +392,50 @@ public void onFailure(Throwable e) {
 
         @Override
         protected void doRun() {
+            // shardID added to the request, as a side effect of calling resolveRequest
             if (checkBlocks() == false) {
                 return;
             }
-            final ShardIterator shardIt = shards(observer.observedState(), internalRequest);
-            final ShardRouting primary = resolvePrimary(shardIt);
-            if (primary == null) {
-                retryBecauseUnavailable(shardIt.shardId(), "No active shards.");
+            final IndexRoutingTable indexRoutingTable = observer.observedState().getRoutingTable().index(request.shardId().getIndex());
+            if (indexRoutingTable == null) {
+                logger.trace("index for shard [{}] is not found, scheduling a retry.", request.shardId());
+                retryBecauseUnavailable(request.shardId(), "index is not active");
```

Comment:
cool. That shards method is really a bad name (I know it was there). maybe rename to shardRoutingTable?